### PR TITLE
[C10D] Document destroy_process_group usage

### LINKED ISSUE #48203: Document torch.distributed.destroy_process_group()
## 📚 Documentation

This is a useful function to de-init a PG in order to re-initialize it, for example during error handling/retries for distributed training. However, the function is not documented in the docs currently: https://pytorch.org/docs/master/distributed.html.

cc @jlin27 @mruberry @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122358

This API was not documented. It has already been a source of confusion,
but recently has become more urgent as improper destruction can lead to
hangs due to ncclCommAbort's requirement of being called collectively.
image

Fixes #48203

### CODE REVIEW

File: docs/source/distributed.rst [hunk 1]
```diff
@@ -293,6 +293,33 @@ check whether the process group has already been initialized use :func:`torch.di
...
+
+It is important to clean up resources on exit by calling :func:`destroy_process_group`.
+
+The simplest pattern to follow is to destroy every process group and backend by calling
+:func:`destroy_process_group()` with the default value of None for the `group` argument, at a
+point in the training script where communications are no longer needed, usually near the
+end of main().  The call should be made once per trainer-process, not at the outer
+process-launcher level.
+
+if :func:`destroy_process_group` is not called, especially when there are multiple process-groups
```
Comment (shuqiangzhang) [inline docs/source/distributed.rst:307]: maybe add something like " if :func:`destroy_process_group` is not called or not called  by all ranks in a pg within the timeout duration". 

we saw some cases in which some ranks (usually rank 0) calls abort just too 'late'.

File: docs/source/distributed.rst [hunk 2]
```diff
@@ -293,6 +293,33 @@ check whether the process group has already been initialized use :func:`torch.di
...
+in the application e.g. for N-D parallelism, hangs on exit are possible.  This is because the
+destructor for ProcessGroupNCCL calls ncclCommAbort, which must be called collectively, but
```
Comment (shuqiangzhang) [inline docs/source/distributed.rst:309]: "destructor for ProcessGroupNCCL calls ncclCommAbort" -> "destructor for ProcessGroupNCCL would call ncclCommAbort if this NCCL API has NOT been called by destroy_process_group before destruction"

**Review by shuqiangzhang — APPROVED** [review]
Some minor addition

**Comment by bhack** [comment]
is this ready to be merged?

**Comment by bhack** [comment]
There is also a document use case in this tutorial https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#basic-use-case

**Comment by wconstab** [comment]
Oops I forgot I had not merged this.

**Pull request closed without merging**